cpp/src/arrow/util/io_util_test.cc

```diff
@@ -77,6 +77,188 @@ TEST(PlatformFilename, Invalid) {
   ASSERT_RAISES(Invalid, PlatformFilename::FromString(s, &fn));
 }
 
+TEST(PlatformFilename, Join) {
+  PlatformFilename fn, joined;
+  ASSERT_OK(PlatformFilename::FromString("a/b", &fn));
+  ASSERT_OK(fn.Join("c/d", &joined));
+  ASSERT_EQ(joined.ToString(), "a/b/c/d");
+#if _WIN32
+  ASSERT_EQ(joined.ToNative(), L"a\\b\\c\\d");
+#else
+  ASSERT_EQ(joined.ToNative(), "a/b/c/d");
+#endif
+
+  ASSERT_OK(PlatformFilename::FromString("a/b/", &fn));
+  ASSERT_OK(fn.Join("c/d", &joined));
+  ASSERT_EQ(joined.ToString(), "a/b/c/d");
+#if _WIN32
+  ASSERT_EQ(joined.ToNative(), L"a\\b\\c\\d");
```

Comment:
I'd really prefer if we could minimize the usage of `#if` like this. Could you instead define a helper which converts to the appropriate character type and joins segments with the platform's separator:
```suggestion
  ASSERT_EQ(joined.ToNative(), NativeJoin({"a", "b", "c", "d"}));
```

Reply:
or even
```suggestion
  ASSERT_EQ(joined.ToNative(), FirstUnlessWindows("a/b/c/d", L"a\\b\\c\\d"));
```

Reply:
I'll create a follow up JIRA for this.

Reply:
Ok, thanks. I like the explicit spelling, because it's clear what we are testing against.